aidial_adapter_bedrock/llm/converse/input.py

```diff
@@ -0,0 +1,325 @@
+    message: DialMessage,
+    storage: FileStorage | None,
+    supported_image_types: list[str] | None = None,
+) -> ConverseMessage:
+
+    return {
+        "role": to_converse_role(message.role),
+        "content": await _get_converse_message_content(
+            message, storage, supported_image_types
+        ),
+    }
+
+
+@dataclass
+class ExtractSystemPromptResult:
+    system_prompt: ConverseTextPart | None
+    system_message_count: int
+    non_system_messages: List[DialMessage]
+
+
+def extract_converse_system_prompt(
+    messages: List[DialMessage],
+) -> ExtractSystemPromptResult:
+    system_msgs = []
+    found_non_system = False
+    system_messages_count = 0
+    non_system_messages = []
+
+    for msg in messages:
+        if msg.role == DialRole.SYSTEM:
+            if found_non_system:
+                raise ValidationError(
+                    "A system message can only follow another system message"
+                )
+            system_messages_count += 1
+            match msg.content:
+                case str():
+                    system_msgs.append(msg.content)
+                case list():
+                    for part in msg.content:
+                        match part:
+                            case MessageContentTextPart():
+                                system_msgs.append(part.text)
+                            case MessageContentImagePart():
+                                raise ValidationError(
+                                    "System messages cannot contain images"
+                                )
+                case None:
+                    pass
+                case _:
+                    assert_never(msg.content)
+        else:
+            found_non_system = True
+            non_system_messages.append(msg)
+    combined = "\n\n".join(msg for msg in system_msgs if msg)
+    return ExtractSystemPromptResult(
+        system_prompt=ConverseTextPart(text=combined) if combined else None,
+        system_message_count=system_messages_count,
+        non_system_messages=non_system_messages,
+    )
+
+
+async def to_converse_messages(
+    messages: List[DialMessage],
+    storage: FileStorage | None,
+    # Offset for system messages at the beginning
+    start_offset: int = 0,
+) -> ListProjection[ConverseMessage]:
+    def _merge(
+        a: Tuple[ConverseMessage, Set[int]],
+        b: Tuple[ConverseMessage, Set[int]],
+    ) -> Tuple[ConverseMessage, Set[int]]:
+        (msg1, set1), (msg2, set2) = a, b
+
+        content1 = msg1["content"]
+        content2 = msg2["content"]
+
+        return {
+            "role": msg1["role"],
+            "content": list(content1) + list(content2),
+        }, set1 | set2
+
+    converted: List[Tuple[ConverseMessage, Set[int]]] = [
+        (await to_converse_message(msg, storage), set([idx + start_offset]))
+        for idx, msg in enumerate(messages)
+    ]
+
+    # Merge messages with same roles, to preserve turn-based user/assistant turns
```

Comment:
# Merge messages with the same roles to achieve an alternation of user-assistant roles.

Reply:
Fixed